### PR TITLE
Implement route profiles in pyvrp.search

### LINKED ISSUE #475: Vehicle-dependent distances and durations
#### Is your feature request related to a problem? Please describe

Right now PyVRP assumes there is a single distance and duration matrix that applies to all vehicle types. This is unrealistic for heterogeneous fleets.

#### Describe the solution you'd like

A way to set different distances and durations for different vehicle types.

#### Additional context

Brought up in #441 by @Stanpol.

### PR DESCRIPTION
**Pull request by N-Wouda**
This PR:

- [x] Closes #475.
- [x] Adds and passes relevant tests.
- [x] Has well-formatted code and documentation.

### Benchmarks

On VRPTW, mostly to double check this has not caused any regressions.

Current main (c61337a4b35ad98be45cb121c905a2feb7629d69; job ID 10298970; 1h / 10 seeds):
```
               MEAN      MIN       MAX   COV
      set
obj   C1    41297.8  40544.2   41410.3   0.6
      C2    16202.5  16197.3   16207.6   0.0
      R1    47307.6  47264.9   47361.5   0.1
      R2    27823.7  27797.3   27858.1   0.1
      RC1   44303.4  44260.9   44372.4   0.1
      RC2   23302.3  23272.8   23339.8   0.1
iters C1    75055.8  60972.4   84106.6  10.6
      C2   104728.4  83505.6  130069.8  15.8
      R1    36980.3  29988.7   45109.0  15.2
      R2    66671.3  54805.1   78681.4  14.7
      RC1   38034.6  29556.4   44465.5  15.4
      RC2   71589.0  56209.1   86459.6  16.8
```
(weirdly variable; probably just some cluster fluke)

This PR (d92ec5dd3e7e41a1590e2d6f6d547ccb3c14f394; job ID 10350889; 1h / 10 seeds):
```
               MEAN      MIN       MAX   COV
      set
obj   C1    41302.4  40613.7   41393.9   0.6
      C2    16201.7  16197.3   16206.9   0.0
      R1    47310.6  47236.8   47397.5   0.1
      R2    27821.7  27806.0   27845.4   0.1
      RC1   44303.8  44271.3   44351.2   0.1
      RC2   23299.8  23277.5   23324.6   0.1
iters C1    69758.4  58638.0   82371.0  10.0
      C2   102283.4  83356.1  125512.8  12.8
      R1    36418.5  29372.6   41728.1  11.4
      R2    65805.7  53118.3   74326.3  11.4
      RC1   37538.6  30485.2   42079.4  10.9
      RC2   69553.8  56292.7   77951.8  11.1
```

<details>

**Notes**:

- It is essential that you add a test when making code changes.
  This keeps the code coverage level up, and helps ensure the changes work as intended.
- When fixing a bug, you must add a test that would produce the bug in the master branch, and then show that it is fixed with the new code. 
- New code additions must be well formatted. Changes should pass the pre-commit workflow, which you can set up locally using [pre-commit](https://pre-commit.com/#intro). 
- Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

**Comment by N-Wouda** [comment]
This should now be functional but I haven't really tested it yet. Tests and stuff are probably for tomorrow. And then we'll need to figure out how well it works with a benchmark or two.

**Comment by N-Wouda** [comment]
I want to add one more high-level test using the `Model` interface. I need to go now, though, so that's for later today. And then this PR should be complete!

**Comment by N-Wouda** [comment]
@leonlan could you have a look at this? I'll have a go at renaming the concatenation schemes in a follow up PR.

**Review by leonlan — APPROVED** [review]

File: tests/search/test_SwapTails.py [hunk 1]
```diff
@@ -172,3 +172,36 @@ def test_move_involving_multiple_depots():
...
+    delta = dist[3, 2] + dist[2, 0] - dist[3, 0] - route2.distance()
+    assert_equal(op.evaluate(route1[1], route2[0], cost_eval), delta)
+
+    # This move evaluates the setting where the first route would be empty, and
+    # the second 0 -> 2 -> 3 -> 0.
+    assert_equal(route1.distance(), 3_994)
+
+    dist = data.distance_matrix(1)
+    delta = dist[2, 3] + dist[3, 0] - dist[2, 0] - route1.distance()
+    assert_equal(op.evaluate(route1[0], route2[1], cost_eval), delta)
```
[the remark this answers is not in the document]
Comment (N-Wouda) [reply tests/search/test_SwapTails.py:207]: How about:
```python
def test_move_with_different_profiles(ok_small_two_profiles):
    """
    Tests that SwapTails correctly evaluates moves between routes with
    different profiles.
    """
    data = ok_small_two_profiles
    dist1, dist2 = data.distance_matrices()

    route1 = Route(data, idx=0, vehicle_type=0)
    route1.append(Node(loc=3))
    route1.update()

    route2 = Route(data, idx=1, vehicle_type=1)
    route2.append(Node(loc=2))
    route2.update()

    op = SwapTails(data)
    cost_eval = CostEvaluator(0, 0, 0)  # all zero so no costs from penalties

    # First route has profile 0, and its distance is thus computed using the
    # first distance matrix.
    assert_equal(route1.profile(), 0)
    assert_equal(route1.distance(), dist1[0, 3] + dist1[3, 0])

    # Second route has profile 1, and its distance is thus computed using the
    # second distance matrix.
    assert_equal(route2.profile(), 1)
    assert_equal(route2.distance(), dist2[0, 2] + dist2[2, 0])

    # This move evaluates the setting where the second route would be empty,
    # and the first 0 -> 3 -> 2 -> 0.
    delta = dist1[3, 2] + dist1[2, 0] - dist1[3, 0] - route2.distance()
    assert_equal(op.evaluate(route1[1], route2[0], cost_eval), delta)

    # This move evaluates the setting where the first route would be empty, and
    # the second 0 -> 2 -> 3 -> 0.
    delta = dist2[2, 3] + dist2[3, 0] - dist2[2, 0] - route1.distance()
    assert_equal(op.evaluate(route1[0], route2[1], cost_eval), delta)
```
?

[the remark this answers is not in the document]
Comment (leonlan) [reply tests/search/test_SwapTails.py]: That also works for me!

**Pull request merged**